In [1]:
#https://towardsdatascience.com/simple-and-multiple-linear-regression-in-python-c928425168f9

import pandas as pd



In [2]:
#df = pd.read_csv('/home/alp/Google Drive/Classes/IC230TS/Excel_HW/election-context-2018.csv')
df = pd.read_csv('G:\My Drive\Classes\IC230TS\Excel_HW\election-context-2018.csv')

In [3]:
df.head(3)

state     county   fips PctDT PctHC PctOth16 PctMR PctBO PctOth12  \
0  Illinois      Adams  17001  71.3  24.0      4.6  66.7  31.5      1.8   
1  Illinois  Alexander  17003  52.6  44.3      3.1  42.5  56.1      1.4   
2  Illinois       Bond  17005  64.9  27.5      7.6  55.8  41.2      3.0   

  Pct>18yo  ... Unnamed: 30 Unnamed: 31 Unnamed: 32 Unnamed: 33 Unnamed: 34  \
0     76.8  ...         NaN         NaN         NaN         NaN         NaN   
1     75.7  ...         NaN         NaN         NaN         NaN         NaN   
2     78.5  ...         NaN         NaN         NaN         NaN         NaN   

  Unnamed: 35 Unnamed: 36 Unnamed: 37 Unnamed: 38 Unnamed: 39  
0         NaN         NaN         NaN         NaN         NaN  
1         NaN         NaN         NaN         NaN         NaN  
2         NaN         NaN         NaN         NaN         NaN  

[3 rows x 40 columns]

In [4]:
df.tail()

state county fips PctDT PctHC PctOth16 PctMR PctBO PctOth12 Pct>18yo  ...  \
127   NaN    NaN  NaN   NaN   NaN      NaN   NaN   NaN      NaN      NaN  ...   
128   NaN    NaN  NaN   NaN   NaN      NaN   NaN   NaN      NaN      NaN  ...   
129   NaN    NaN  NaN   NaN   NaN      NaN   NaN   NaN      NaN      NaN  ...   
130   NaN    NaN  NaN   NaN   NaN      NaN   NaN   NaN      NaN      NaN  ...   
131   NaN    NaN  NaN   NaN   NaN      NaN   NaN   NaN      NaN      NaN  ...   

    Unnamed: 30 Unnamed: 31 Unnamed: 32 Unnamed: 33 Unnamed: 34 Unnamed: 35  \
127         NaN         NaN         NaN         NaN         NaN         NaN   
128         NaN         NaN         NaN         NaN         NaN         NaN   
129         NaN         NaN         NaN         NaN         NaN         NaN   
130         NaN         NaN         NaN         NaN         NaN         NaN   
131         NaN         NaN         NaN         NaN         NaN         NaN   

    Unnamed: 36 Unnamed: 37 Unnamed: 38 Unnamed: 39  
127         NaN         NaN         NaN         NaN  
128         NaN         NaN         NaN         NaN  
129         NaN         NaN         NaN         NaN  
130         NaN         NaN         NaN         NaN  
131         NaN         NaN         NaN         NaN  

[5 rows x 40 columns]

Need to clean up the data: drop all these NaN rows.

In [5]:
df.columns

Index(['state', 'county', 'fips', 'PctDT', 'PctHC', 'PctOth16', 'PctMR',
       'PctBO', 'PctOth12', 'Pct>18yo', 'PctTurnout', 'white_pct', 'black_pct',
       'hispanic_pct', 'nonwhite_pct', 'foreignborn_pct', 'female_pct',
       'age29andunder_pct', 'age65andolder_pct', 'median_hh_inc',
       'clf_unemploy_pct', 'lesshs_pct', 'lesscollege_pct',
       'lesshs_whites_pct', 'lesscollege_whites_pct', 'rural_pct',
       'ruralurban_cc', 'Unnamed: 27', 'Unnamed: 28', 'Unnamed: 29',
       'Unnamed: 30', 'Unnamed: 31', 'Unnamed: 32', 'Unnamed: 33',
       'Unnamed: 34', 'Unnamed: 35', 'Unnamed: 36', 'Unnamed: 37',
       'Unnamed: 38', 'Unnamed: 39'],
      dtype='object')

In [3]:
df= df[['state', 'county', 'fips', 'PctDT', 'PctHC', 'PctOth16', 'PctMR',
       'PctBO', 'PctOth12', 'Pct>18yo', 'PctTurnout', 'white_pct', 'black_pct',
       'hispanic_pct', 'nonwhite_pct', 'foreignborn_pct', 'female_pct',
       'age29andunder_pct', 'age65andolder_pct', 'median_hh_inc',
       'clf_unemploy_pct', 'lesshs_pct', 'lesscollege_pct',
       'lesshs_whites_pct', 'lesscollege_whites_pct', 'rural_pct',
       'ruralurban_cc']]
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132 entries, 0 to 131
Data columns (total 27 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   state                   102 non-null    object 
 1   county                  125 non-null    object 
 2   fips                    126 non-null    object 
 3   PctDT                   125 non-null    object 
 4   PctHC                   124 non-null    object 
 5   PctOth16                123 non-null    object 
 6   PctMR                   122 non-null    object 
 7   PctBO                   121 non-null    object 
 8   PctOth12                120 non-null    object 
 9   Pct>18yo                119 non-null    object 
 10  PctTurnout              118 non-null    object 
 11  white_pct               117 non-null    object 
 12  black_pct               116 non-null    object 
 13  hispanic_pct            115 non-null    object 
 14  nonwhite_pct            114 non-null    ob

And need to change teh "object" (string) to numbers (floats)

In [4]:
df = df.dropna()

In [8]:
df.tail()

state      county   fips PctDT PctHC PctOth16 PctMR PctBO PctOth12  \
97   Illinois   Whiteside  17195  49.3  43.1      7.6  40.3  57.8      1.9   
98   Illinois        Will  17197  44.1  50.5      5.4  46.5  52.0      1.5   
99   Illinois  Williamson  17199  67.7  26.9      5.3  61.5  36.5      2.0   
100  Illinois   Winnebago  17201  47.0  47.1      5.9  46.3  51.9      1.8   
101  Illinois    Woodford  17203  66.5  25.6      7.9  68.7  29.5      1.8   

    Pct>18yo  ... age29andunder_pct age65andolder_pct median_hh_inc  \
97      76.1  ...              36.0              18.9         49151   
98      68.3  ...              41.2              11.2         77507   
99      77.2  ...              36.2              17.5         45902   
100     72.3  ...              38.7              15.6         49468   
101     74.8  ...              38.4              16.3         68040   

    clf_unemploy_pct lesshs_pct lesscollege_pct lesshs_whites_pct  \
97               6.8       11.3            82.7               9.7   
98               6.8        9.2            66.5               4.8   
99               7.6        9.8            77.9               9.6   
100             10.5       12.6            77.6               8.8   
101              5.3        6.0            69.6               5.6   

    lesscollege_whites_pct rural_pct ruralurban_cc  
97                    81.9      37.3           4.0  
98                    64.3       3.9           1.0  
99                    78.0      34.8           3.0  
100                   75.2       7.9           2.0  
101                   69.5      54.9           2.0  

[5 rows x 27 columns]

In [5]:
col_list = df.columns
col_list[3:]

Index(['PctDT', 'PctHC', 'PctOth16', 'PctMR', 'PctBO', 'PctOth12', 'Pct>18yo',
       'PctTurnout', 'white_pct', 'black_pct', 'hispanic_pct', 'nonwhite_pct',
       'foreignborn_pct', 'female_pct', 'age29andunder_pct',
       'age65andolder_pct', 'median_hh_inc', 'clf_unemploy_pct', 'lesshs_pct',
       'lesscollege_pct', 'lesshs_whites_pct', 'lesscollege_whites_pct',
       'rural_pct', 'ruralurban_cc'],
      dtype='object')

In [6]:
for col in col_list[3:]:
    df[col] = df[col].astype(float)
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 102 entries, 0 to 101
Data columns (total 27 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   state                   102 non-null    object 
 1   county                  102 non-null    object 
 2   fips                    102 non-null    object 
 3   PctDT                   102 non-null    float64
 4   PctHC                   102 non-null    float64
 5   PctOth16                102 non-null    float64
 6   PctMR                   102 non-null    float64
 7   PctBO                   102 non-null    float64
 8   PctOth12                102 non-null    float64
 9   Pct>18yo                102 non-null    float64
 10  PctTurnout              102 non-null    float64
 11  white_pct               102 non-null    float64
 12  black_pct               102 non-null    float64
 13  hispanic_pct            102 non-null    float64
 14  nonwhite_pct            102 non-null    fl

OK, data should be clean. Let's sample a few counties to check.

In [10]:
df.sample(5)

state   county   fips  PctDT  PctHC  PctOth16  PctMR  PctBO  PctOth12  \
65  Illinois   Mercer  17131   56.4   36.0       7.6   45.4   52.8       1.8   
23  Illinois  Edwards  17047   83.0   13.0       4.0   74.5   23.4       2.1   
31  Illinois   Grundy  17063   57.9   34.7       7.4   53.4   44.5       2.1   
59  Illinois  Madison  17119   54.1   38.9       7.0   49.5   48.1       2.4   
43  Illinois  Johnson  17087   76.4   18.8       4.9   69.9   27.7       2.4   

    Pct>18yo  ...  age29andunder_pct  age65andolder_pct  median_hh_inc  \
65      77.9  ...               33.8               20.3        55750.0   
23      76.8  ...               35.2               19.4        47035.0   
31      73.1  ...               39.8               12.9        67162.0   
59      76.7  ...               38.1               15.6        54573.0   
43      81.3  ...               33.9               19.0        44179.0   

    clf_unemploy_pct  lesshs_pct  lesscollege_pct  lesshs_whites_pct  \
65               6.1         8.1             82.5                8.0   
23               7.2        11.0             88.0               10.8   
31               7.5         7.1             78.2                5.9   
59               8.2         7.4             74.3                7.0   
43               8.9        16.6             83.3               11.7   

    lesscollege_whites_pct  rural_pct  ruralurban_cc  
65                    82.6       77.8            2.0  
23                    88.3      100.0            9.0  
31                    78.1       24.4            1.0  
59                    73.6       13.3            1.0  
43                    81.2      100.0            8.0  

[5 rows x 27 columns]

Data looks OK. 

Let's look at a correlation table for the IL county data. . . 

In [8]:
corr1 = df.corr()
corr1.style.background_gradient(cmap='coolwarm',axis=None).format(precision=3)

/tmp/ipykernel_27125/550126231.py:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  corr1 = df.corr()


Let's set up a regression analysis to see which of these variables is most important in determining pct vote for DT.

In [7]:
import statsmodels.api as sm

In [10]:
y = df['PctDT'] # dependent
x = df['PctMR'] # independent
x = sm.add_constant(x)

model = sm.OLS(y,x).fit()
predictions = model.predict(x)

print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                  PctDT   R-squared:                       0.801
Model:                            OLS   Adj. R-squared:                  0.799
Method:                 Least Squares   F-statistic:                     401.6
Date:                Mon, 06 Feb 2023   Prob (F-statistic):           8.52e-37
Time:                        21:01:33   Log-Likelihood:                -316.10
No. Observations:                 102   AIC:                             636.2
Df Residuals:                     100   BIC:                             641.4
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -1.6683      3.259     -0.512      0.6

PctMR coeff means that for every point support for Romney (2012) increased, there was a 1.13 increase for support of Trump (2016).

So our regression equation is $ PctDT  =  -1.668 + 1.1315 * PctMR $

$ R^2 = 0.801 $ means that this equation explains about 80% of support for Trump based on support for Romney.

How does this compare with support for HC by Obama?

In [11]:
y = df['PctHC']
x = df['PctBO']
x = sm.add_constant(x)

model = sm.OLS(y,x).fit()
predictions = model.predict(x)

print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                  PctHC   R-squared:                       0.822
Model:                            OLS   Adj. R-squared:                  0.820
Method:                 Least Squares   F-statistic:                     461.1
Date:                Mon, 06 Feb 2023   Prob (F-statistic):           3.10e-39
Time:                        21:01:52   Log-Likelihood:                -303.71
No. Observations:                 102   AIC:                             611.4
Df Residuals:                     100   BIC:                             616.7
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        -12.5320      2.093     -5.988      0.0

It seems like support for Romney was a stronger predictor for support for DT (1.13) than support for Obama was for HC (1.0669). (I think.)

Let's try some multivariate regressions. . . 

If I take variables that are positively correlated with support for DT, what sort of MLR can we get?

In [15]:
df.columns

Index(['state', 'county', 'fips', 'PctDT', 'PctHC', 'PctOth16', 'PctMR',
       'PctBO', 'PctOth12', 'Pct>18yo', 'PctTurnout', 'white_pct', 'black_pct',
       'hispanic_pct', 'nonwhite_pct', 'foreignborn_pct', 'female_pct',
       'age29andunder_pct', 'age65andolder_pct', 'median_hh_inc',
       'clf_unemploy_pct', 'lesshs_pct', 'lesscollege_pct',
       'lesshs_whites_pct', 'lesscollege_whites_pct', 'rural_pct',
       'ruralurban_cc'],
      dtype='object')

In [12]:
y = df['PctDT'] # dependent varible
x = df[['white_pct','age65andolder_pct','median_hh_inc','lesshs_pct', 'lesscollege_pct','rural_pct','ruralurban_cc' ]] # independent variable(s)
x = sm.add_constant(x)

model = sm.OLS(y,x).fit()
predictions = model.predict(x)

print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                  PctDT   R-squared:                       0.803
Model:                            OLS   Adj. R-squared:                  0.788
Method:                 Least Squares   F-statistic:                     54.72
Date:                Mon, 06 Feb 2023   Prob (F-statistic):           2.04e-30
Time:                        21:02:04   Log-Likelihood:                -315.50
No. Observations:                 102   AIC:                             647.0
Df Residuals:                      94   BIC:                             668.0
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const               -36.2066     11.42

So these variables -- which are positively correlated with voting for DT -- achieve an $R^2$ of 0.803.

Let's look at a different set of variables -- those that are positively correclated with HC -- to see what happens.

In [13]:
y = df['PctHC'] # dependent varible
x = df[['black_pct', 'hispanic_pct', 'nonwhite_pct', 'female_pct',
       'age29andunder_pct',  'median_hh_inc','lesshs_pct', 'lesscollege_pct',
       'rural_pct', 'ruralurban_cc']]
x = sm.add_constant(x)

model = sm.OLS(y,x).fit()
predictions = model.predict(x)


print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                  PctHC   R-squared:                       0.838
Model:                            OLS   Adj. R-squared:                  0.820
Method:                 Least Squares   F-statistic:                     47.01
Date:                Mon, 06 Feb 2023   Prob (F-statistic):           1.26e-31
Time:                        21:02:17   Log-Likelihood:                -298.90
No. Observations:                 102   AIC:                             619.8
Df Residuals:                      91   BIC:                             648.7
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                85.5516     25.20

What happens if we use ALL the demographic variables? Do we increase the $R^2$?

In [14]:
y = df['PctHC'] # dependent varible
x = df[['white_pct', 'black_pct',
       'hispanic_pct', 'nonwhite_pct', 'foreignborn_pct', 'female_pct',
       'age29andunder_pct', 'age65andolder_pct', 'median_hh_inc',
       'clf_unemploy_pct', 'lesshs_pct', 'lesscollege_pct',
       'lesshs_whites_pct', 'lesscollege_whites_pct', 'rural_pct',
       'ruralurban_cc']]
x = sm.add_constant(x)

model = sm.OLS(y,x).fit()
predictions = model.predict(x)

print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                  PctHC   R-squared:                       0.847
Model:                            OLS   Adj. R-squared:                  0.821
Method:                 Least Squares   F-statistic:                     31.78
Date:                Mon, 06 Feb 2023   Prob (F-statistic):           1.10e-28
Time:                        21:02:22   Log-Likelihood:                -295.87
No. Observations:                 102   AIC:                             623.7
Df Residuals:                      86   BIC:                             665.7
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
const                      0

So we go from $R^2$ of 83.8 to 86.4. A bit better, but not much. . . 

In [15]:
y = df['PctDT'] # dependent varible
x = df[['Pct>18yo', 'PctTurnout', 'white_pct', 'black_pct',
       'hispanic_pct', 'nonwhite_pct', 'foreignborn_pct', 'female_pct',
       'age29andunder_pct', 'age65andolder_pct', 'median_hh_inc',
       'clf_unemploy_pct', 'lesshs_pct', 'lesscollege_pct',
       'lesshs_whites_pct', 'lesscollege_whites_pct', 'rural_pct',
       'ruralurban_cc']]
x = sm.add_constant(x)

model = sm.OLS(y,x).fit()
predictions = model.predict(x)

print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                  PctDT   R-squared:                       0.855
Model:                            OLS   Adj. R-squared:                  0.825
Method:                 Least Squares   F-statistic:                     29.08
Date:                Mon, 06 Feb 2023   Prob (F-statistic):           4.41e-28
Time:                        21:02:41   Log-Likelihood:                -299.94
No. Observations:                 102   AIC:                             635.9
Df Residuals:                      84   BIC:                             683.1
Df Model:                          17                                         
Covariance Type:            nonrobust                                         
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
const                      0

When we do the calculation for DT, we go from $R^2$ of 80.3 to 85.5. This is a bigger gain than for HC, but our first DT MLR didn't have as many variables, so it's hard to compare these.

Now, let's get the national data (which will include the 102 counties in IL) and see how the same variables work on the larger data set.

In [8]:
import pandas as pd
nation = pd.read_excel('G:\My Drive\Classes\IC230TS\Excel_HW\election-context-2018.xlsx',engine='openpyxl')
nation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3114 entries, 0 to 3113
Data columns (total 43 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   state                   3114 non-null   object 
 1   county                  3114 non-null   object 
 2   fips                    3114 non-null   int64  
 3   trump16                 3114 non-null   int64  
 4   pctDT                   3113 non-null   float64
 5   clinton16               3114 non-null   int64  
 6   pctHC                   3113 non-null   float64
 7   otherpres16             3114 non-null   int64  
 8   total16                 3114 non-null   int64  
 9   romney12                3114 non-null   int64  
 10  obama12                 3114 non-null   int64  
 11  otherpres12             3114 non-null   int64  
 12  total12                 3114 non-null   int64  
 13  demsen16                1942 non-null   float64
 14  repsen16                1942 non-null   

In [9]:
nation = nation.fillna(0)

In [13]:
nation.columns

Index(['state', 'county', 'fips', 'trump16', 'pctDT', 'clinton16', 'pctHC',
       'otherpres16', 'total16', 'romney12', 'obama12', 'otherpres12',
       'total12', 'demsen16', 'repsen16', 'othersen16', 'demhouse16',
       'rephouse16', 'otherhouse16', 'demgov16', 'repgov16', 'othergov16',
       'repgov14', 'demgov14', 'othergov14', 'total_population', 'cvap',
       'white_pct', 'black_pct', 'hispanic_pct', 'nonwhite_pct',
       'foreignborn_pct', 'female_pct', 'age29andunder_pct',
       'age65andolder_pct', 'median_hh_inc', 'clf_unemploy_pct', 'lesshs_pct',
       'lesscollege_pct', 'lesshs_whites_pct', 'lesscollege_whites_pct',
       'rural_pct', 'ruralurban_cc'],
      dtype='object')

In [14]:
nation.sample(5)

state    county   fips  trump16      pctDT  clinton16      pctHC  \
230    Colorado     Delta   8029    11655  69.416319       4087  24.341870   
452     Georgia  McIntosh  13191     3487  59.071658       2303  39.014061   
1105  Louisiana    Iberia  22045    20903  64.408085      10698  32.963579   
1815   New York  Franklin  36033     8221  48.495753       7297  43.045068   
472     Georgia      Polk  13233    11014  77.481534       2867  20.168836   

      otherpres16  total16  romney12  ...  age29andunder_pct  \
230          1048    16790     10915  ...          32.358294   
452           113     5903      3409  ...          30.801234   
1105          853    32454     20892  ...          42.332552   
1815         1434    16952      5231  ...          36.947870   
472           334    14215      9811  ...          41.342122   

      age65andolder_pct  median_hh_inc  clf_unemploy_pct  lesshs_pct  \
230           23.324840        42011.0          9.153575   11.989151   
452           20.120508        45248.0          9.929773   17.604197   
1105          13.599100        44573.0         11.173695   22.209812   
1815          14.929323        49782.0          8.549105   14.641301   
472           14.880159        39356.0          8.730637   24.288622   

      lesscollege_pct  lesshs_whites_pct  lesscollege_whites_pct  rural_pct  \
230         80.393509           7.985870               78.295777  63.172008   
452         86.233362          19.032550               83.167571  74.311030   
1105        86.255585          18.606939               83.203253  28.136264   
1815        81.598503          12.170530               80.182710  62.675633   
472         87.015726          20.202939               86.229838  51.421338   

      ruralurban_cc  
230             6.0  
452             3.0  
1105            2.0  
1815            7.0  
472             4.0  

[5 rows x 43 columns]

First, DT on the smaller set of variables.

In [18]:
y = nation['pctDT'] # dependent varible
x = nation[['white_pct','age65andolder_pct','median_hh_inc','lesshs_pct', 'lesscollege_pct','rural_pct','ruralurban_cc' ]] # independent variable(s)
x = sm.add_constant(x)

model = sm.OLS(y,x).fit()
predictions = model.predict(x)

print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                  pctDT   R-squared:                       0.569
Model:                            OLS   Adj. R-squared:                  0.568
Method:                 Least Squares   F-statistic:                     585.1
Date:                Mon, 06 Feb 2023   Prob (F-statistic):               0.00
Time:                        21:03:03   Log-Likelihood:                -11698.
No. Observations:                3114   AIC:                         2.341e+04
Df Residuals:                    3106   BIC:                         2.346e+04
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const               -43.6468      2.87

In [10]:
y = nation['pctHC'] # dependent varible
x = nation[['black_pct', 'hispanic_pct', 'nonwhite_pct', 'female_pct',
       'age29andunder_pct',  'median_hh_inc','lesshs_pct', 'lesscollege_pct',
       'rural_pct', 'ruralurban_cc']]
x = sm.add_constant(x) 

model = sm.OLS(y,x).fit()
predictions = model.predict(x)

print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                  pctHC   R-squared:                       0.626
Model:                            OLS   Adj. R-squared:                  0.625
Method:                 Least Squares   F-statistic:                     518.9
Date:                Sun, 19 Feb 2023   Prob (F-statistic):               0.00
Time:                        09:55:45   Log-Likelihood:                -11380.
No. Observations:                3114   AIC:                         2.278e+04
Df Residuals:                    3103   BIC:                         2.285e+04
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                80.4495      3.91

Now, HC for ALL the variables. Does this move the $R^2$ from 0.626, which we got with the smaller set of variables?

In [11]:
y = nation['pctHC'] # dependent varible

x = nation[['white_pct', 'black_pct',
       'hispanic_pct', 'nonwhite_pct', 'foreignborn_pct', 'female_pct',
       'age29andunder_pct', 'age65andolder_pct', 'median_hh_inc',
       'clf_unemploy_pct', 'lesshs_pct', 'lesscollege_pct',
       'lesshs_whites_pct', 'lesscollege_whites_pct', 'rural_pct',
       'ruralurban_cc']]

x = sm.add_constant(x)

model = sm.OLS(y,x).fit()
predictions = model.predict(x)

print(model.summary())



                            OLS Regression Results                            
Dep. Variable:                  pctHC   R-squared:                       0.656
Model:                            OLS   Adj. R-squared:                  0.654
Method:                 Least Squares   F-statistic:                     369.5
Date:                Sun, 19 Feb 2023   Prob (F-statistic):               0.00
Time:                        09:56:13   Log-Likelihood:                -11248.
No. Observations:                3114   AIC:                         2.253e+04
Df Residuals:                    3097   BIC:                         2.263e+04
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
const                     56

With all the variables, the $R^2$ increases a bit (to 0.656) but not much. There is still a lot of variation in the voting that these variables don't explain.

Let's look at the correlation tables for the national data and the IL only data.

In [21]:
# National Data (including IL)
corr1 = nation[['pctDT', 'pctHC','white_pct', 'black_pct', 'hispanic_pct', 'nonwhite_pct',
       'foreignborn_pct', 'female_pct', 'age29andunder_pct',
       'age65andolder_pct', 'median_hh_inc', 'clf_unemploy_pct', 'lesshs_pct',
       'lesscollege_pct', 'lesshs_whites_pct', 'lesscollege_whites_pct',
       'rural_pct', 'ruralurban_cc']].corr()
corr1.style.background_gradient(cmap='coolwarm',axis=None).format(precision=3)

In [22]:
# Illinois Only
corr1 = df[['PctDT', 'PctHC','white_pct', 'black_pct', 'hispanic_pct', 'nonwhite_pct',
       'foreignborn_pct', 'female_pct', 'age29andunder_pct',
       'age65andolder_pct', 'median_hh_inc', 'clf_unemploy_pct', 'lesshs_pct',
       'lesscollege_pct', 'lesshs_whites_pct', 'lesscollege_whites_pct',
       'rural_pct', 'ruralurban_cc']].corr()
corr1.style.background_gradient(cmap='coolwarm',axis=None).format(precision=3)

After typing all the stuff below, I found a simple explaination here: https://medium.com/swlh/interpreting-linear-regression-through-statsmodels-summary-4796d359035a

### Durbin-Watson statistic.
https://www.statsmodels.org/stable/generated/statsmodels.stats.stattools.durbin_watson.html#statsmodels.stats.stattools.durbin_watson
* The null hypothesis of the test is that there is no serial correlation in the residuals.
* The test statistic is approximately equal to 2*(1-r) where r is the sample autocorrelation of the residuals. Thus, for r == 0, indicating no serial correlation, the test statistic equals 2. This statistic will always be between 0 and 4. The closer to 0 the statistic, the more evidence for positive serial correlation. The closer to 4, the more evidence for negative serial correlation.

### Omnibus test for normality. (= _Chi_^2 test)


### kurtosis measures
https://www.simplypsychology.org/kurtosis.html
* Kurtosis is a statistical measure used to describe the degree to which scores cluster in the tails or the peak of a frequency distribution. The peak is the tallest part of the distribution, and the tails are the ends of the distribution.

* There are three types of kurtosis: mesokurtic, leptokurtic, and platykurtic.
 * When kurtosis is equal to 3, the distribution is mesokurtic. This means the kurtosis is the same as the normal distribution, it is mesokurtic (medium peak). The kurtosis of a mesokurtic distribution is neither high nor low, rather it is considered to be a baseline for the two other classifications.
 * Positive excess values of kurtosis (>3) indicate that a distribution is peaked and possess thick tails. Leptokurtic distributions have positive kurtosis values. A leptokurtic distribution has a higher peak (thin bell) and taller (i.e. fatter and heavy) tails than a normal distribution. 
 * Negative excess values of kurtosis (<3) indicate that a distribution is flat and has thin tails.
 * Platykurtic distributions have negative kurtosis values. A platykurtic distribution is flatter (less peaked) when compared with the normal distribution, with fewer values in its shorter (i.e. lighter and thinner) tails.

### Skewness.
https://www.statisticshowto.com/probability-and-statistics/descriptive-statistics/skewness/
* Skewness is a measure of symmetry in a distribution. Actually, it’s more correct to describe it as a measure of lack of symmetry. A standard normal distribution is perfectly symmetrical and has zero skew.
 * A symmetrical distribution has a skew of zero. 
 * A positive result means that your data is positively skewed. (That is, right-tail.)
 * A negative results means that your data is negatively skewed. (Left-tail.)

### Jarque-Bera test of normality.
https://www.statsmodels.org/stable/generated/statsmodels.stats.stattools.jarque_bera.html#statsmodels.stats.stattools.jarque_bera
https://en.wikipedia.org/wiki/Jarque%E2%80%93Bera_test
* The Jarque–Bera test is a goodness-of-fit test of whether sample data have the skewness and kurtosis matching a normal distribution. Samples from a normal distribution have an expected skewness of 0 and an expected excess kurtosis of 0 (which is the same as a kurtosis of 3).
* 


In [12]:
# let's explore how the different states generate different values . . .

states = nation['state'].value_counts()

In [13]:
states.index

Index(['Texas', 'Georgia', 'Virginia', 'Kentucky', 'Missouri', 'Kansas',
       'Illinois', 'North Carolina', 'Iowa', 'Tennessee', 'Nebraska',
       'Indiana', 'Ohio', 'Minnesota', 'Michigan', 'Mississippi', 'Oklahoma',
       'Arkansas', 'Wisconsin', 'Alabama', 'Pennsylvania', 'Florida',
       'South Dakota', 'Colorado', 'Louisiana', 'New York', 'California',
       'Montana', 'West Virginia', 'North Dakota', 'South Carolina', 'Idaho',
       'Washington', 'Oregon', 'New Mexico', 'Utah', 'Maryland', 'Wyoming',
       'New Jersey', 'Nevada', 'Maine', 'Arizona', 'Vermont', 'Massachusetts',
       'New Hampshire', 'Connecticut', 'Rhode Island', 'Hawaii', 'Delaware',
       'District of Columbia'],
      dtype='object')

In [14]:
y = nation['pctHC'] # dependent varible

x = nation[['white_pct', 'black_pct',
       'hispanic_pct', 'female_pct',
       'age29andunder_pct', 'age65andolder_pct', 'median_hh_inc',
       'clf_unemploy_pct', 'lesshs_pct', 'lesscollege_pct',
        'rural_pct']]

x = sm.add_constant(x)

model = sm.OLS(y,x).fit()
predictions = model.predict(x)

print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                  pctHC   R-squared:                       0.633
Model:                            OLS   Adj. R-squared:                  0.632
Method:                 Least Squares   F-statistic:                     486.9
Date:                Sun, 19 Feb 2023   Prob (F-statistic):               0.00
Time:                        09:57:11   Log-Likelihood:                -11349.
No. Observations:                3114   AIC:                         2.272e+04
Df Residuals:                    3102   BIC:                         2.279e+04
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                93.7878      4.20

In [21]:
y = nation['pctDT'] # dependent varible

x = nation[['white_pct', 'black_pct',
       'hispanic_pct', 'female_pct',
       'age29andunder_pct', 'age65andolder_pct', 'median_hh_inc',
       'clf_unemploy_pct', 'lesshs_pct', 'lesscollege_pct',
        'rural_pct']]

x = sm.add_constant(x)

model = sm.OLS(y,x).fit()
predictions = model.predict(x)

print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                  pctDT   R-squared:                       0.615
Model:                            OLS   Adj. R-squared:                  0.614
Method:                 Least Squares   F-statistic:                     450.7
Date:                Fri, 17 Feb 2023   Prob (F-statistic):               0.00
Time:                        16:37:31   Log-Likelihood:                -11521.
No. Observations:                3114   AIC:                         2.307e+04
Df Residuals:                    3102   BIC:                         2.314e+04
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const               -23.1597      4.44

In [22]:
model.rsquared

0.6151194204842272

In [23]:
model.params

const               -23.159663
white_pct             0.358576
black_pct            -0.121019
hispanic_pct         -0.010571
female_pct           -0.433512
age29andunder_pct     0.373050
age65andolder_pct     0.321412
median_hh_inc         0.000089
clf_unemploy_pct     -0.963269
lesshs_pct            0.523850
lesscollege_pct       0.672927
rural_pct             0.063668
dtype: float64

In [15]:
state_rsquare = {}

for s in states.index:
    subset = nation[nation['state'] == s]
    if len(subset) < 15:
        print(f"Model cannot run because {s} has only {len(subset)} counties.")
        #state_rsquare[s] = -999
    else:

        y = subset['pctHC'] # dependent varible

        x = subset[['white_pct', 'black_pct',
        'hispanic_pct', 'female_pct',
        'age29andunder_pct', 'age65andolder_pct', 'median_hh_inc',
        'clf_unemploy_pct', 'lesshs_pct', 'lesscollege_pct',
            'rural_pct']]

        x = sm.add_constant(x)

        model = sm.OLS(y,x).fit()

        state_rsquare[s] = [len(subset),model.rsquared,model.params]


Model cannot run because Vermont has only 14 counties.
Model cannot run because Massachusetts has only 14 counties.
Model cannot run because New Hampshire has only 10 counties.
Model cannot run because Connecticut has only 8 counties.
Model cannot run because Rhode Island has only 5 counties.
Model cannot run because Hawaii has only 4 counties.
Model cannot run because Delaware has only 3 counties.
Model cannot run because District of Columbia has only 1 counties.


In [16]:
state_rsquare['Illinois']

[102,
 0.830272049778888,
 const                51.394324
 white_pct            -0.276477
 black_pct             0.463932
 hispanic_pct          0.824694
 female_pct            0.336087
 age29andunder_pct     0.123948
 age65andolder_pct     1.158618
 median_hh_inc        -0.000023
 clf_unemploy_pct      0.392038
 lesshs_pct           -0.794643
 lesscollege_pct      -0.432280
 rural_pct            -0.052297
 dtype: float64]

So this is what my dictionary looks like: key = state name; value = list of numbers (n counties, $r^2$, then regression values).

How do I access these? Let's mess around a bit and make sure I've got the syntax correct.

In [17]:
state_rsquare['Texas'][1]

0.8013710695434562

In [24]:
from heapq import nlargest

res = nlargest(10, state_rsquare, key = lambda state_rsquare : state_rsquare[1] )

print(f"{'State':<14}{'N counties':>10}{'R-squared':>10}{'white':>10}{'black':>10}{'hispanic':>10}{'female':>10}{'<29YO':>10}{'>65YO':>10}{'hh_inc':>10}{'unemploy':>10}{'<hs':>10}{'<college':>10}{'rural':>10}")

for item in res:    
    coef_str = ''
    for i in range(1,12):
        coef_str += f"{state_rsquare[item][2][i]:10.3}"
    print(f"{str(item):<15}{state_rsquare[item][0]:10}{state_rsquare[item][1]:10.5}{coef_str}")

State         N counties R-squared     white     black  hispanic    female     <29YO     >65YO    hh_inc  unemploy       <hs  <college     rural
Wyoming                23   0.95561    -0.586   0.00884     0.122     -0.28     -1.01     -1.31 -0.000278      0.12    -0.189     -1.05 -0.000812
Utah                   29   0.93528    -0.746      1.61   -0.0842    -0.714     -2.25      -2.2 -0.000227      0.61    -0.655    -0.536   -0.0613
Arkansas               75   0.89698    -0.723    -0.135    -0.714 -0.000972    -0.282   -0.0548 -0.000108     0.108     0.498    -0.672   -0.0415
Oregon                 36   0.86981    -0.703     -1.68     0.151      3.65     -2.13     -2.57 -0.000681      -1.0    -0.498     -1.02    -0.109
Arizona                15   0.94482     -1.17     -5.77    -0.681      -9.3     -5.65     -5.04  -0.00187     -2.75      2.34     -2.97    -0.441
North Carolina        100   0.92779    -0.296     0.493    -0.143    0.0629    -0.149    -0.275 -0.000464    -0.355  -0.00166

The `.nlargest()` function isn't working propertly. I'm trying to sort on the $R^2$ values and get the top (and bottom) ten. Iowa shouldn't be listed (0.59) here. And the results are not actually sorted in top-to-bottom order anyway!

Let's try just sorting the results.

In [18]:
results = sorted(state_rsquare.items(), key = lambda x : x[1][1], reverse= True)[:10] # state
#for item in results:
#    print(item,state_rsquare[item][1])

In [29]:
type(results)

list

In [28]:
results[0]

('Nevada',
 [17, 0.9886893408675105, const                193.615892
  white_pct             -0.420362
  black_pct              0.019346
  hispanic_pct           0.591271
  female_pct             0.515286
  age29andunder_pct     -1.592152
  age65andolder_pct     -0.726917
  median_hh_inc         -0.000142
  clf_unemploy_pct       0.859871
  lesshs_pct             0.451775
  lesscollege_pct       -1.295604
  rural_pct             -0.041053
  dtype: float64])

In [25]:
for item in results[0:2]:
    print(item[1][1]) # r values

0.9886893408675105
0.9781048298066375


In [27]:
for item in results[0:2]:
    print(item[1][2]) # interesting -- the regression figures are a list of values, so we'll extract them by going item[1][2][x]

const                193.615892
white_pct             -0.420362
black_pct              0.019346
hispanic_pct           0.591271
female_pct             0.515286
age29andunder_pct     -1.592152
age65andolder_pct     -0.726917
median_hh_inc         -0.000142
clf_unemploy_pct       0.859871
lesshs_pct             0.451775
lesscollege_pct       -1.295604
rural_pct             -0.041053
dtype: float64
const                93.717956
white_pct            -0.674736
black_pct             0.199012
hispanic_pct         -0.290176
female_pct            0.611941
age29andunder_pct    -0.284620
age65andolder_pct    -0.341180
median_hh_inc        -0.000203
clf_unemploy_pct     -0.123606
lesshs_pct            0.186307
lesscollege_pct      -0.351290
rural_pct            -0.012090
dtype: float64


In [31]:
results = sorted(state_rsquare.items(), key = lambda x : x[1][1], reverse= True)[:10]

print(f"{'State':<15}{'N counties':>10}{'R-squared':>10}{'white':>10}{'black':>10}{'hispanic':>10}{'female':>10}{'<29YO':>10}{'>65YO':>10}{'hh_inc':>10}{'unemploy':>10}{'<hs':>10}{'<college':>10}{'rural':>10}")

for item in results:
    state = item[0]
    counties = item[1][0] # N counties
    r_squared = item[1][1] 
    coef_str = ''
    for i in range(1,12):
        coef_str += f"{item[1][2][i]:10.3}"
    print(f"{str(state):<15}{counties:10}{r_squared:10.3}{coef_str}")


State          N counties R-squared     white     black  hispanic    female     <29YO     >65YO    hh_inc  unemploy       <hs  <college     rural
Nevada                 17     0.989     -0.42    0.0193     0.591     0.515     -1.59    -0.727 -0.000142      0.86     0.452      -1.3   -0.0411
Alabama                67     0.978    -0.675     0.199     -0.29     0.612    -0.285    -0.341 -0.000203    -0.124     0.186    -0.351   -0.0121
Mississippi            82     0.976   -0.0857      0.84     0.643    -0.207     0.761     0.977  0.000128   0.00536     0.112    -0.165   0.00575
South Carolina         46     0.971    -0.884   -0.0477     -0.58    -0.256   -0.0932    -0.197 -0.000319   -0.0511    -0.387    -0.374   -0.0206
Maryland               24      0.97    -0.217     0.584     0.722     0.633    -0.157    -0.194 -7.95e-05     0.505    -0.432    -0.626   0.00802
Louisiana              64     0.968    -0.617     0.407     0.172      1.15    -0.899     -1.26   5.1e-05     0.059      0.2

In [32]:
results = sorted(state_rsquare.items(), key = lambda x : x[1][1], reverse= True)[-10:]

print(f"{'State':<15}{'N counties':>10}{'R-squared':>10}{'white':>10}{'black':>10}{'hispanic':>10}{'female':>10}{'<29YO':>10}{'>65YO':>10}{'hh_inc':>10}{'unemploy':>10}{'<hs':>10}{'<college':>10}{'rural':>10}")

for item in results:
    state = item[0]
    counties = item[1][0] # N counties
    r_squared = item[1][1] 
    coef_str = ''
    for i in range(1,12):
        coef_str += f"{item[1][2][i]:10.3}"
    print(f"{str(state):<15}{counties:10}{r_squared:10.3}{coef_str}")


State          N counties R-squared     white     black  hispanic    female     <29YO     >65YO    hh_inc  unemploy       <hs  <college     rural
Washington             39     0.776    -0.397     0.372   -0.0675     0.729     -0.39     0.398  0.000446      2.04    0.0344    -0.896   -0.0294
Kansas                105      0.77    -0.597     0.481    -0.648     0.555    -0.541     -1.05 -0.000186     0.778     0.205    -0.556   -0.0539
South Dakota           66      0.76    -0.326     0.979     0.607    -0.601     0.457     0.266  6.25e-05     0.237    0.0226    -0.397     -0.03
Minnesota              87     0.757    -0.424     0.998     0.348     -1.09    -0.017     0.622 -0.000237     0.432     -1.39    -0.544   -0.0357
Wisconsin              72     0.706    -0.313      0.43   -0.0349    -0.645    -0.383     -1.74 -0.000854    -0.585    -0.802    -0.809    0.0897
West Virginia          55     0.696     -1.26     -1.11    -0.311     0.444   -0.0101     0.192 -2.23e-05    0.0694     0.20

In [38]:
state = 'Texas'
subset = nation[nation['state'] == state]
y = subset['pctHC'] # dependent varible

x = subset[['white_pct', 'black_pct','hispanic_pct', 'female_pct',
        'age29andunder_pct', 'age65andolder_pct', 'median_hh_inc',
        'clf_unemploy_pct', 'lesshs_pct', 'lesscollege_pct',
            'rural_pct']]

x = sm.add_constant(x)

model = sm.OLS(y,x).fit()

#predictions = model.predict(x)

print(f"Results for {state}. N counties: {len(subset)}")
print(model.summary())

Results for Texas. N counties: 254
                            OLS Regression Results                            
Dep. Variable:                  pctHC   R-squared:                       0.801
Model:                            OLS   Adj. R-squared:                  0.792
Method:                 Least Squares   F-statistic:                     88.76
Date:                Sun, 19 Feb 2023   Prob (F-statistic):           2.15e-78
Time:                        10:34:01   Log-Likelihood:                -857.86
No. Observations:                 254   AIC:                             1740.
Df Residuals:                     242   BIC:                             1782.
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
con

In [6]:
string = 'able was i ere i saw elba'
if string[::1] == string[::-1]:
    print ('palindrome')
else:
    print ('not palindrome')

palindrome
